### PR TITLE
Discriminative training script for WSJ/TDNN-LSTM

### PR DESCRIPTION
**Pull request by hhadian**
I have not tested this script (I don't have any WSJ models). Do you want me to test it?

### CODE REVIEW

**Comment by danpovey** [comment]
yes, test it please; you can link to stuff in /home/dpovey/kald-trunk for
the earlier parts of the system.

**Comment by hhadian** [comment]
OK will do.

**Comment by hhadian** [comment]
This is ready. The results:
```
# System                tdnn_lstm1a_sp_smbr:1 tdnn_lstm1a_sp_smbr:2 tdnn_lstm1a_sp_smbr:3
#WER dev93 (tgpr)                8.09      8.03      8.02
#             [looped:]          8.02      7.98      8.00
#WER dev93 (tg)                  7.86      7.85      7.81
#             [looped:]          7.69      7.69      7.63
#WER dev93 (big-dict,tgpr)       5.84      5.83      5.87
#             [looped:]          5.78      5.82      5.94
#WER dev93 (big-dict,fg)         5.39      5.39      5.31
#             [looped:]          5.34      5.34      5.31
#WER eval92 (tgpr)               6.15      6.10      6.01
#             [looped:]          6.10      6.08      6.06
#WER eval92 (tg)                 5.72      5.76      5.65
#             [looped:]          5.85      5.76      5.72
#WER eval92 (big-dict,tgpr)      3.67      3.51      3.53
#             [looped:]          3.69      3.51      3.49
#WER eval92 (big-dict,fg)        3.05      3.17      3.10
#             [looped:]          3.19      3.24      3.17
```

**Comment by danpovey** [comment]
i wonder if the learning rate is too low (if data is smaller we generally
increase it).
can you show some objf values?

**Comment by hhadian** [comment]
These are the last 15 valid objf's:
```
objf_valid.50.log:LOG 0.978801
objf_valid.51.log:LOG 0.978355
objf_valid.52.log:LOG 0.978695
objf_valid.53.log:LOG 0.978822
objf_valid.54.log:LOG 0.978567
objf_valid.55.log:LOG 0.978625
objf_valid.56.log:LOG 0.979239
objf_valid.57.log:LOG 0.979311
objf_valid.58.log:LOG 0.978995
objf_valid.59.log:LOG 0.978664
objf_valid.60.log:LOG 0.979045
objf_valid.61.log:LOG 0.978849
objf_valid.62.log:LOG 0.978868
objf_valid.63.log:LOG 0.978948
objf_valid.64.log:LOG 0.979034
objf_valid.65.log:LOG 0.979076
```
What learning rate do you suggest I try?

**Comment by danpovey** [comment]
those don't tell me much as they are too close together.  I need them from
the start.

**Comment by hhadian** [comment]
This is from the start (but not all of them):
```
objf_valid.0.log:LOG 0.974206
objf_valid.3.log:LOG 0.974994
objf_valid.5.log:LOG 0.975437
objf_valid.9.log:LOG 0.976588
objf_valid.10.log:LOG 0.976028
objf_valid.15.log:LOG 0.977002
objf_valid.20.log:LOG 0.977584
objf_valid.25.log:LOG 0.977803
objf_valid.30.log:LOG 0.978066
objf_valid.35.log:LOG 0.97905
objf_valid.40.log:LOG 0.978357
objf_valid.45.log:LOG 0.979115
objf_valid.50.log:LOG 0.978801
objf_valid.55.log:LOG 0.978625
objf_valid.60.log:LOG 0.979045
objf_valid.65.log:LOG 0.979076
```
Do you want me to send all of them? I tried to generate a plot but it seems it failed.

**Comment by danpovey** [comment]
I think it's changing too slowly.  Try twice or 4 times that learning rate
(you can run 2 separate experiments).
However the training objfs would have been more diagnostic.

**Comment by hhadian** [comment]
OK, will try it.
And this is from train:
```
objf_train.0.log:LOG 0.971025
objf_train.3.log:LOG 0.973752
objf_train.5.log:LOG 0.973866
objf_train.9.log:LOG 0.974524
objf_train.10.log:LOG 0.974534
objf_train.15.log:LOG 0.975687
objf_train.20.log:LOG 0.977415
objf_train.25.log:LOG 0.978772
objf_train.30.log:LOG 0.978685
objf_train.35.log:LOG 0.979589
objf_train.40.log:LOG 0.979932
objf_train.45.log:LOG 0.981117
objf_train.50.log:LOG 0.982086
objf_train.55.log:LOG 0.981997
objf_train.60.log:LOG 0.982178
objf_train.65.log:LOG 0.982952
```

**Comment by danpovey** [comment]
just try twice the learning rate, I think 4 times may be too much.

**Comment by hhadian** [comment]
OK

**Comment by hhadian** [comment]
This is with 2x larger learning rate (i.e. 0.0000050):
```
# System                tdnn_lstm1a_sp_smbr_1b:1 tdnn_lstm1a_sp_smbr_1b:2 tdnn_lstm1a_sp_smbr_1b:3
#WER dev93 (tgpr)                8.04      8.14      8.29
#             [looped:]          7.99      8.14      8.36
#WER dev93 (tg)                  7.78      7.81      7.93
#             [looped:]          7.72      7.83      7.93
#WER dev93 (big-dict,tgpr)       5.76      5.96      6.02
#             [looped:]          5.89      5.98      6.05
#WER dev93 (big-dict,fg)         5.36      5.47      5.32
#             [looped:]          5.31      5.37      5.42
#WER eval92 (tgpr)               6.06      6.01      6.01
#             [looped:]          6.04      5.99      6.03
#WER eval92 (tg)                 5.67      5.56      5.53
#             [looped:]          5.69      5.64      5.74
#WER eval92 (big-dict,tgpr)      3.60      3.56      3.58
#             [looped:]          3.58      3.60      3.65
#WER eval92 (big-dict,fg)        3.10      3.10      3.14
#             [looped:]          3.23      3.19      3.31
```
And these are some training objf values:
```
objf_train.0.log:LOG 0.971025
objf_train.3.log:LOG 0.974453
objf_train.5.log:LOG 0.974937
objf_train.9.log:LOG 0.976403
objf_train.10.log:LOG 0.976461
objf_train.15.log:LOG 0.978229
objf_train.20.log:LOG 0.980491
objf_train.25.log:LOG 0.981819
objf_train.30.log:LOG 0.981797
objf_train.35.log:LOG 0.982642
objf_train.40.log:LOG 0.98333
objf_train.45.log:LOG 0.984626
objf_train.50.log:LOG 0.985371
objf_train.55.log:LOG 0.985161
objf_train.60.log:LOG 0.985022
objf_train.65.log:LOG 0.986363
```

**Comment by danpovey** [comment]
hm, maybe it's too fast now. maybe the original one was better.   compare
also with the baseline. == epoch0.

**Comment by hhadian** [comment]
This is the original one compared to baseline (tdnn_lstm1a_sp):
```
# System                tdnn_lstm1a_sp tdnn_lstm1a_sp_smbr:1 tdnn_lstm1a_sp_smbr:2 tdnn_lstm1a_sp_smbr:3
#WER dev93 (tgpr)                8.54      8.09      8.03      8.02
#             [looped:]          8.54      8.02      7.98      8.00
#WER dev93 (tg)                  8.25      7.86      7.85      7.81
#             [looped:]          8.21      7.69      7.69      7.63
#WER dev93 (big-dict,tgpr)       6.24      5.84      5.83      5.87
#             [looped:]          6.28      5.78      5.82      5.94
#WER dev93 (big-dict,fg)         5.70      5.39      5.39      5.31
#             [looped:]          5.70      5.34      5.34      5.31
#WER eval92 (tgpr)               6.52      6.15      6.10      6.01
#             [looped:]          6.45      6.10      6.08      6.06
#WER eval92 (tg)                 6.13      5.72      5.76      5.65
#             [looped:]          6.08      5.85      5.76      5.72
#WER eval92 (big-dict,tgpr)      3.88      3.67      3.51      3.53
#             [looped:]          3.93      3.69      3.51      3.49
#WER eval92 (big-dict,fg)        3.38      3.05      3.17      3.10
#             [looped:]          3.47      3.19      3.24      3.17
```

**Comment by danpovey** [comment]
ok original LR was OK then.  so it's ready to merge?

**Comment by hhadian** [comment]
Yes, but there is a conflict related to the kernel. I need to resolve it.
